doc/design/concurrent_data_transfer_and_kernel_execution.md

```diff
@@ -0,0 +1,97 @@
+  
+<p align="center"><img src="images/gpu_async.png" height="75%" width = "75%" align="center"/><br/></p>
+  
+## Solution
+One feasible method is to use [staging area](https://en.wikipedia.org/wiki/Staging_(data)). A staging area is an intermediate storage area used for data processing during the [extract, transform and load (ETL)](https://en.wikipedia.org/wiki/Extract,_transform,_load) process. The data staging area sits between the data source(s) and the data target(s), which are often data warehouses, data marts, or other data repositories.   
+The staging area has one fixed-size buffer and two methods, `put` and `get`. `put` and `get` access buffer in a thread-safe way. The operation of `put` puts data into the buffer, and `get` removes a data from the buffer. When the program runs, the staging area should be warmed up. At first, `put` is called to put some data in the staging area. After that, `get` is called when the kernel begins executing, and one data is extracted from the staging area. At the same time, `put` is called and a new data is copied from CPU side to the staging area. Because the operation of data transfer is asynchronous, so the `put` method will return immediately.     
+
+The GPU provided by NVIDIA generally has two engines: a copy engine and a computing engine. To make two engines working at the same time, we need to use the [stream mechanism](https://devblogs.nvidia.com/parallelforall/gpu-pro-tip-cuda-7-streams-simplify-concurrency/) provided by CUDA. As long as data transfer and kernel execution are in different streams and the kernel does not use the data being transferred, the data transfer and kernel execution can be done at the same time on GPU.  
+
+To support the above description, we need to define a new class: `Channel`. Here, we borrow the concept of [`Channel`](https://www.golang-book.com/books/intro/10#section2) in the go language.
+
+```
+template <typename T>
+class Channel {
+ private:
+  using ChannelElement = std::vector<T>;
+
+  std::size_t capacity_;
+  std::size_t bytes_limit_;
+  std::size_t current_bytes_;
+  std::mutex mu_;
+  std::condition_variable empty_cond_var_;
+  std::condition_variable full_cond_var_;
+  std::deque<ChannelElement> channel_;
+
+ public:
+  explicit Channel(std::size_t capacity, std::size_t bytes_limit);
+
+  void Put(ChannelElement* buffer_element) { ... }
+  void Get(ChannelElement* buffer_element) { ... }
+  size_t Size() { ... }
+  void Clear()  { ... }
+};
+};
+```
+
+**Take the [recognize_digits_mlp](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_recognize_digits_mlp.py) as an example to show the model definition after using the staging area mechanism.**
+
+```
+...
+concurrent_program = Program()
```

Comment:
Do we need to expose the channel concept to the user?
We can do the following, and use channel as the implementation, without exposing it to the user:
```
main_program = Program()
with program(main_program):
    image = fluid.layers.data_layer(..., buf_size=100)
    label = fluid.layers.data_layer(..., buf_size=100)
    with parallel.for(places) as for_loop:
        y_predict = fluid.layers.fc(input=image, size=1, act=None)
        cost = fluid.layers.square_error_cost(input=y_predict, label=label)
        avg_cost = fluid.layers.mean(x=cost)
        ....

for i in range(buffer_size):
    data = next(train_reader())
    executor.run(concurrent_program, feed=feeder.feed(data))

for pass_id in range(PASS):
    for data in train_reader():
        executor.run(concurrent_program, feed=[...])
        executor.run(main_program, fetch=[...])

for i in range(buffer_size):
    executor.run(main_program, fetch=[...])
```

Please compare the Python code and the above code, feels like exposing the channel concept only adds verbosity to the program.

Reply:
> Do we need to expose the channel concept to the user?

I think we should have a deeper discussion about this.

> automatically create a separate thread for data loading.

Currently, data loading is done on Python side, and we do not consider this in this PR.
This PR mainly analyzes the concurrent execution of data transferring from the CPU end to GPU transmission and computing model at the GPU.

> feels like exposing the channel concept to the user only adds verbosity to the program

Exposing the channel concept maybe not appropriate, and some user may never use `golang`. But this can make the program more flexible.